_posts/2022-11-25-async-sink-rate-limiting-strategy.md

```diff
@@ -0,0 +1,195 @@
+
+First, we implement the information methods to keep track of the number of in-flight requests and in-flight messages.
+
+```java
+public class CongestionControlRateLimitingStrategy implements RateLimitingStrategy {
+    // ...
+    @Override
+    public void registerInFlightRequest(RequestInfo requestInfo) {
+        currentInFlightRequests++;
+        currentInFlightMessages += requestInfo.getBatchSize();
+    }
+    
+    @Override
+    public void registerCompletedRequest(ResultInfo resultInfo) {
+        currentInFlightRequests = Math.max(0, currentInFlightRequests - 1);
+        currentInFlightMessages = Math.max(0, currentInFlightMessages - resultInfo.getBatchSize());
+        
+        if (resultInfo.getFailedMessages() > 0) {
+            maxInFlightMessages = scalingStrategy.scaleDown(maxInFlightMessages);
+        } else {
+            maxInFlightMessages = scalingStrategy.scaleUp(maxInFlightMessages);
+        }
+    }
+    // ...
+}
+```
+
+Then we implement the control methods to dynamically adjust the request rate.
+
+We keep a current value for maxInFlightMessages and maxInFlightRequests, and postpone all new requests if maxInFlightRequests or maxInFlightMessages have been reached.
+
+Every time a request completes, the CongestionControlRateLimitingStrategy will check if there are any failed messages in the response. If there are, it will decrease maxInFlightMessages. If there are no failed messages, it will increase maxInFlightMessages. This gives us an indirect control of rate of messages being written to the destination.
+
+Side note: The default CongestionControlRateLimitingStrategy uses an Additive Increase / Multiplicative Decrease (AIMD) scaling strategy. This is also used in TCP congestion control to avoid overloading the destination by increasing write rate slowly but backing off quickly if throttled.
+
+```java
+public class CongestionControlRateLimitingStrategy implements RateLimitingStrategy {
+    // ...
+    @Override
+    public void registerCompletedRequest(ResultInfo resultInfo) {
+        // ...
+        if (resultInfo.getFailedMessages() > 0) {
+            maxInFlightMessages = scalingStrategy.scaleDown(maxInFlightMessages);
+        } else {
+            maxInFlightMessages = scalingStrategy.scaleUp(maxInFlightMessages);
+        }
+    }
+    
+    public boolean shouldBlock(RequestInfo requestInfo) {
+        return currentInFlightRequests >= maxInFlightRequests
+                || (currentInFlightMessages + requestInfo.getBatchSize() > maxInFlightMessages);
+    }
+    // ...
+}
+```
+
+### [Example 2] TokenBucketRateLimitingStrategy
+
+The CongestionControlRateLimitingStrategy is rather aggressive, and relies on a robust server-side rate limiting strategy. In the event we don’t have a robust server-side rate limiting strategy, we can implement a client-side rate limiting strategy.
```

Comment:
Is the TokenBucketRateLimitingStrategy also part of Flink?

Reply:
Nope, we did not make it a part of Flink as part of our initial implementation. If people find this strategy useful, I can create a JIRA for this, it would be quite simple to add!

Reply:
I can not assess if people would find it useful or not. Maybe it makes sense to just create a JIRA to that people can comment on it, and or upvote it; without actually adding the code for now.